src/array/dyn_geometry_array.rs

```diff
@@ -0,0 +1,310 @@
+use crate::array::zip_validity::ZipValidity;
+use crate::array::{
+    GeometryCollectionArray, LineStringArray, MixedGeometryArray, MultiLineStringArray,
+    MultiPointArray, MultiPolygonArray, PointArray, PolygonArray, RectArray, WKBArray,
+};
+use crate::datatypes::GeoDataType;
+use crate::error::GeoArrowError;
+use crate::scalar::Geometry;
+use crate::trait_::GeometryArrayAccessor;
+use crate::GeometryArrayTrait;
+use arrow_array::{Array, OffsetSizeTrait};
+use arrow_buffer::bit_iterator::BitIterator;
+use arrow_schema::Field;
+
+impl dyn GeometryArrayTrait {
+    pub fn value<O: OffsetSizeTrait>(&self, i: usize) -> Geometry<O> {
```

Comment:
This `O` seems problematic, as it would require you to specify the offset type in advance, right? But you don't know the type in advance until you look at the `data_type`.

In general, I don't think this method should exist, or at least should hardly ever be used, because it would require downcasting on every element instead of once per array, which is contrary to my perf goals.

For algorithms, you should downcast _once_ for the entire array to a strongly-typed array, and then not need dynamic dispatch for the iteration within the array.

Reply:
So impl algorithms for `dyn GeometryArrayTrait` like this?
```rust
impl BoundingRect for dyn GeometryArrayTrait {
    fn bounding_rect(&self) -> RectArray {
        match self.data_type() {
            GeoDataType::Point(_) => {
                let array = as_point_array(self);
                let output_geoms: Vec<Option<Rect>> = array
                    .iter_geo()
                    .map(|maybe_g| maybe_g.map(|geom| geom.bounding_rect()))
                    .collect();

                RectArray::from(output_geoms)
            }
            GeoDataType::LineString(_) => {
                let array = as_line_string_array::<i32>(self);
                let output_geoms: Vec<Option<Rect>> = array
                    .iter_geo()
                    .map(|maybe_g| maybe_g.and_then(|geom| geom.bounding_rect()))
                    .collect();

                RectArray::from(output_geoms)
            }
            _ => unimplemented!()
        }
    }
}
```

Reply:
I think we should be able to extend the existing macro to handle this. I started some scratch work in https://github.com/geoarrow/geoarrow-rs/pull/326.

Given that `BoundingRect` is implemented for the concrete arrays already, we don't need to repeat the implementation of casting to geo and calling its `bounding_rect` method.

That can simplify to

```rs
impl BoundingRect for dyn GeometryArrayTrait {
    fn bounding_rect(&self) -> RectArray {
        match self.data_type() {
            GeoDataType::Point(_) => as_point_array(self).bounding_rect(),
            GeoDataType::LineString(_) => as_line_string_array::<i32>(self).bounding_rect(),
            _ => unimplemented!()
        }
    }
}
```